eng/GenerateAnalyzerNuspec.targets

```diff
@@ -88,7 +88,7 @@
       <AnalyzerRulesetAssembly Condition="'@(AnalyzerNupkgAssembly)' != '' and '@(AnalyzerRulesetAssembly)' == ''" Include="@(AnalyzerNupkgAssembly)"/>
       <AnalyzerRulesetAssembly Update="@(AnalyzerRulesetAssembly)" Condition="'%(AnalyzerRulesetAssembly.TargetFramework)' == ''" TargetFramework="$(TargetFramework)" />
     </ItemGroup>
-    <Exec Command='"$(DotNetRoot)dotnet.exe" "$(_GenerateDocumentationAndConfigFilesPath)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorconfigsDir)" "$(ArtifactsBinDir)\" "$(Configuration)" "%(AnalyzerRulesetAssembly.TargetFramework)" "@(AnalyzerRulesetAssembly)" "$(PackagePropsFileDir)" "$(PackagePropsFileName)" "$(DisableNETAnalyzersPackagePropsFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerDocumentationFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerSarifFileName)" "$(VersionPrefix)" $(NuspecPackageId) $(ContainsPortedFxCopRules) $(GenerateAnalyzerRulesMissingDocumentationFile)' />
+    <Exec Command='"$(DotNetRoot)dotnet.exe" "$(_GenerateDocumentationAndConfigFilesPath)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorconfigsDir)" "$(ArtifactsBinDir)\" "$(Configuration)" "%(AnalyzerRulesetAssembly.TargetFramework)" "@(AnalyzerRulesetAssembly)" "$(PackagePropsFileDir)" "$(PackagePropsFileName)" "$(DisableNETAnalyzersPackagePropsFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerDocumentationFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerSarifFileName)" "$(VersionPrefix)" $(NuspecPackageId) $(ContainsPortedFxCopRules) $(GenerateAnalyzerRulesMissingDocumentationFile) false' />
```

Comment:
Is this change correct? If so, where/how I should pass true for CI? Will it be written in azure-pipelines.yml? And how I will get the correct paths to pass from azure-pipelines.yml.

I haven't dealt with these infrastructure things before ;-(

Reply:
I would just do an inline Exec, i.e.

```xml
< !-- Validation-only invocation to ensure generated files are not hand edited -->
<Exec Command='"$(DotNetRoot)dotnet.exe" "$(_GenerateDocumentationAndConfigFilesPath)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorconfigsDir)" "$(ArtifactsBinDir)\" "$(Configuration)" "%(AnalyzerRulesetAssembly.TargetFramework)" "@(AnalyzerRulesetAssembly)" "$(PackagePropsFileDir)" "$(PackagePropsFileName)" "$(DisableNETAnalyzersPackagePropsFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerDocumentationFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerSarifFileName)" "$(VersionPrefix)" $(NuspecPackageId) $(ContainsPortedFxCopRules) $(GenerateAnalyzerRulesMissingDocumentationFile) true' />

< !-- Invocation to generate files -->
<Exec Command='"$(DotNetRoot)dotnet.exe" "$(_GenerateDocumentationAndConfigFilesPath)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorconfigsDir)" "$(ArtifactsBinDir)\" "$(Configuration)" "%(AnalyzerRulesetAssembly.TargetFramework)" "@(AnalyzerRulesetAssembly)" "$(PackagePropsFileDir)" "$(PackagePropsFileName)" "$(DisableNETAnalyzersPackagePropsFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerDocumentationFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerSarifFileName)" "$(VersionPrefix)" $(NuspecPackageId) $(ContainsPortedFxCopRules) $(GenerateAnalyzerRulesMissingDocumentationFile) false' />
```

You may also want to move the validation-only bool to be the very first argument so it is easier to read.

Reply:
> You may also want to move the validation-only bool to be the very first argument so it is easier to read.

Or convert it to a string that must start with `-validationOnly:` and end with `true` or `false`, appropriately adjust the parsing.

Reply:
````xml
< !-- Validation-only invocation to ensure generated files are not hand edited -->
<Exec Command='"$(DotNetRoot)dotnet.exe" "$(_GenerateDocumentationAndConfigFilesPath)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorconfigsDir)" "$(ArtifactsBinDir)\" "$(Configuration)" "%(AnalyzerRulesetAssembly.TargetFramework)" "@(AnalyzerRulesetAssembly)" "$(PackagePropsFileDir)" "$(PackagePropsFileName)" "$(DisableNETAnalyzersPackagePropsFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerDocumentationFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerSarifFileName)" "$(VersionPrefix)" $(NuspecPackageId) $(ContainsPortedFxCopRules) $(GenerateAnalyzerRulesMissingDocumentationFile) true' />

< !-- Invocation to generate files -->
<Exec Command='"$(DotNetRoot)dotnet.exe" "$(_GenerateDocumentationAndConfigFilesPath)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorconfigsDir)" "$(ArtifactsBinDir)\" "$(Configuration)" "%(AnalyzerRulesetAssembly.TargetFramework)" "@(AnalyzerRulesetAssembly)" "$(PackagePropsFileDir)" "$(PackagePropsFileName)" "$(DisableNETAnalyzersPackagePropsFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerDocumentationFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerSarifFileName)" "$(VersionPrefix)" $(NuspecPackageId) $(ContainsPortedFxCopRules) $(GenerateAnalyzerRulesMissingDocumentationFile) false' />
```

If this is run outside the CI, wouldn't this fail the build when the files doesn't match the expected ones? or this will fail CI only?

One approach that just came to my mind is to keep using 17 arguments as before. But have some sort of a special environment variable for CI. But not sure which way should we go.

Reply:
> If this is run outside the CI, wouldn't this fail the build when the files doesn't match the expected ones? or this will fail CI only?

Good point. You can condition this on the MSBuild property that is set only on CI? https://github.com/dotnet/roslyn-analyzers/blob/3916933c6691dbd9af67d06f8a49939a8cfdc46a/eng/GenerateAnalyzerNuspec.targets#L43-L45

Reply:
Actually, I think it may be fine if this fails outside CI - if I hand edited the file locally, and run pack, it will fail locally, which seems desirable..

Reply:
@mavasani My concern here is what will happen after it fails locally? will it print the failure message and **continue** executing the generation process? or will it just terminate completely and won't run the generation process?

The failure shouldn't prevent the generation process locally. I'll test and try to find out what happens in this case.

Reply:
> The failure shouldn't prevent the generation process locally.

Why not? The message would indicate which files need to be reverted, so the revert can be done manually by the user?

Reply:
@mavasani I changed the `Main` code to:

```csharp
            const int expectedArguments = 18;

            if (args.Length != expectedArguments)
            {
                Console.Error.WriteLine($"Excepted {expectedArguments} arguments, found {args.Length}: {string.Join(';', args)}");
                return 1;
            }

            if (!bool.TryParse(args[17], out var validateOnly))
            {
                validateOnly = false;
            }

            Console.Error.WriteLine($"VALIDATE ONLY VALUE IS {validateOnly}");
            return validateOnly ? 100 : 200;
```

and had both commands (with validateOnly true and validateOnly false):

```xml
<Exec Command='"$(DotNetRoot)dotnet.exe" "$(_GenerateDocumentationAndConfigFilesPath)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorconfigsDir)" "$(ArtifactsBinDir)\" "$(Configuration)" "%(AnalyzerRulesetAssembly.TargetFramework)" "@(AnalyzerRulesetAssembly)" "$(PackagePropsFileDir)" "$(PackagePropsFileName)" "$(DisableNETAnalyzersPackagePropsFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerDocumentationFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerSarifFileName)" "$(VersionPrefix)" $(NuspecPackageId) $(ContainsPortedFxCopRules) $(GenerateAnalyzerRulesMissingDocumentationFile) true' />
    <Exec Command='"$(DotNetRoot)dotnet.exe" "$(_GenerateDocumentationAndConfigFilesPath)" "$(_GeneratedRulesetsDir)" "$(_GeneratedEditorconfigsDir)" "$(ArtifactsBinDir)\" "$(Configuration)" "%(AnalyzerRulesetAssembly.TargetFramework)" "@(AnalyzerRulesetAssembly)" "$(PackagePropsFileDir)" "$(PackagePropsFileName)" "$(DisableNETAnalyzersPackagePropsFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerDocumentationFileName)" "$(AnalyzerSarifFileDir)" "$(AnalyzerSarifFileName)" "$(VersionPrefix)" $(NuspecPackageId) $(ContainsPortedFxCopRules) $(GenerateAnalyzerRulesMissingDocumentationFile) false' />
```

I only got return code `100` with that. This means that the failure from the first `Exec` command prevents the second `Exec` command from running, which is bad IMO for the reasons stated in https://github.com/dotnet/roslyn-analyzers/pull/4295#discussion_r502683496


---

To double check, I then changed the `Main` code to:

```csharp
            const int expectedArguments = 18;

            if (args.Length != expectedArguments)
            {
                Console.Error.WriteLine($"Excepted {expectedArguments} arguments, found {args.Length}: {string.Join(';', args)}");
                return 1;
            }

            if (!bool.TryParse(args[17], out var validateOnly))
            {
                validateOnly = false;
            }

            return validateOnly ? 0 : 200;
```

and the error code 200 appeared in the logs.

---

To be more clear, the second `Exec` will run **if and only if** the first one succeeds.
In the meanwhile, if the first one succeeds, then running the second one is no different than not running it (it will write the same file contents, so it's the same effect if it's not run at all).

Reply:
A more common scenario is `RulesMissingDocumentation` file. Consider the following:

1. One or more rules currently doesn't have a documentation and they *correctly* exist in the `RulesMissingDocumentation` file.
2. Later, some of these rules were documented.
3. Running `msbuild /t:pack` will now complain about `RulesMissingDocumentation`, without re-generating it.
4. You'll have to manually update RulesMissingDocumentation and remove the documented rules. This step is extra and shouldn't happen. Updating RulesMissingDocumentation should be done automatically as part of running `msbuild /t:pack`

Reply:
@Youssef1313 `RulesMissingDocumentation` files should be excluded from validation. Otherwise, anyone authoring a new doc for a new rule will bring down all of our CI ;)

Reply:
> If a new analyzer is written, the generated files are now different. The user haven't made any manual changes to the files in this case.

Yes, hence I edited my proposed error message in https://github.com/dotnet/roslyn-analyzers/pull/4295#discussion_r502677294. However, note that this validator is likely going to make things very onerous. For example, the auto-generated files include every diagnostic message/title/description. So anyone writing a PR to change any of these will end up seeing CI failures if they just edited, build the project normally and pushed a change. Maybe folks will get used to running pack command in future and it won't be so bad. However, I wanted to call it out that if the validator is indeed adding more burden to the process then helping us, we may have to turn it off.

Reply:
@Youssef1313 Seems we can go to your original proposal of running the validator only on CI by guarding it with the `Sign` MSBuild property being true?

Reply:
@mavasani the validator will most likely require running the pack command in most of PRs. While this adds a little overhead to contributors, it also ensures that generated files are always up to date, which is good.

For `RulesMissingDocumentation`, I'm thinking to be in-between.

- If an expected rule in the file doesn't exist -> fail CI (this happens when adding a new rule)
- If an existing rule in the file wasn't expected -> don't fail CI (this happens when the documentation for a rule is added)

But if you don't like this approach, I'll completely exclude RulesMissingDocumentation from validation.

Reply:
I am fine either way for `RulesMissingDocumentation` files, as long as adding a new rule documentation does not break CI.